### ISSUE TITLE
BUG: 3.2 - DropdownField getSourceAsArray() uses original array key instead of record IDs

### ISSUE THREAD

**Issue by andrewvt** [silverstripe/silverstripe-framework#4841]
@stevie-mayhew - Looks like you added this for 3.2, so hoping you have some context for if this is a good fix.

Problem: validation checks IDs against original array keys and produces "Please select a value within the list provided" message when the ID is not a key.

The problem is this function simply reuses the original keys when re-keying the same array of objects.  

``` php
    // DropdownField.php line 313
    public function getSourceAsArray()
    {
        $source = $this->getSource();
        if (is_array($source)) {
            return $source;
        } else {
            $sourceArray = array();
            foreach ($source as $key => $value) {
                $sourceArray[$key] = $value;
            }
        }
        return $sourceArray;
    }
```

From it's usage for instance in CheckboxSetField line 336 

``` php
    if (!array_intersect_key($sourceArray, $values)) {
```

it is checking IDs against the original array keys instead of IDs against IDs

``` php
        $sourceArray = $this->getSourceAsArray();
    if (is_array($values)) {
        if (!array_intersect_key($sourceArray, $values)) {
            $validator->validationError(
                $this->name,
                _t(
                    'CheckboxSetField.SOURCE_VALIDATION',
                    "Please select a value within the list provided. '{value}' is not a valid option",
                    array('value' => implode(' and ', array_diff($sourceArray, $values)))
                ),
                "validation"
            );
            return false;
        }
    } else {
```

Changing getSourceAsArray() in DropdownField.php as below fixes the problems I was encountering:

``` php
    public function getSourceAsArray()
    {
        $source = $this->getSource();
        if (is_array($source)) {
            return $source;
        } else {
            $sourceArray = array();
            foreach ($source as $key => $value) {
                $sourceArray[$value->ID] = $value;
            }
        }
        return $sourceArray;
    }
```

**Comment by stevie-mayhew** [comment]
@andrewvt that looks reasonable - an over simplification of the behaviour on my part without considering ID sets which aren't 0-based.

Would you be able to do a PR with that change? A test displaying the currently failing behaviour and your fix fixing it would probably be appreciated by the SilverStripe core team.

**Comment by andrewvt** [comment]
Sure, I can work up a pull request and tests.  

In the case of the site we were upgrading and encountered the issue the IDs skipped in a few places so the last two items in our list failed but the rest worked fine since those last ids didn't have corresponding keys.  Was such odd behavior I felt compelled to get to the bottom of it ;)

**Comment by stevie-mayhew** [comment]
Very appreciated! Not worth validating invalid things :smile:

**Comment by nzphoenix** [comment]
Hi @andrewvt any luck with this? I've seemingly encountered the same issue with CheckboxSetField that don't have a numerically indexed ID.

**Comment by mediabeastnz** [comment]
I've across this twice now. @andrewvt how did you get on with a PR?

This seems like a pretty critical thing to fix? It means every last value on a field that uses this base class wont be validated.

**Comment by tractorcow** [comment]
`$sourceArray[$value->ID] = $value;`

Dont' forget, $source could be a non-array that doesn't contain dataobjects. E.g. an SS_Map instance (which might be sourced from a raw sql query).

I suggest to use the `SS_List->map()` method here, since it'll automatically convert objects to arrays safely.

**Comment by amanpilgrim** [comment]
Hi @tractorcow, it seems like the issue has been identified and a logical solution proposed. Are you able to implement the bug fix? I have a CWP client who is affected by this and doesn't understand why they can't update their CheckboxSetField without errors. Thanks

**Comment by tractorcow** [comment]
I think this might have been fixed, but I'll double check later. :)

**Comment by KevinRefinery** [comment]
This in still a bug.

**Comment by andrewvt** [comment]
This whole section got rewritten in a later release of SilverStripe 3 and the issue seemed to go away for us.  Are you saying this is still an issue in 3.2 or are you saying you are having the issue in version 3.5 or higher?

**Comment by KevinRefinery** [comment]
3.5, but only some of the time. I have 2 DataObjects with an enum, one is fine, one isn’t.  Do you want me to zip and send you my project?

**Comment by KevinRefinery** [comment]
Here are some screen shots

**Comment by KevinRefinery** [comment]
The DataObject that has the problem extends Member. Hope that helps.

**Comment by tractorcow** [comment]
Hi @KevinRefinery , sorry your screenshots don't come in via email replies. You'll need to post these on github directly.

**Comment by KevinRefinery** [comment]
The problem for me was a space in the field name.

**Comment by tractorcow** [comment]
I've confirmed the original issue is fixed in 3.5 and above.
